lib/statsd/instrument/udp_sink.rb

```diff
@@ -5,45 +5,72 @@ module Instrument
     # @note This class is part of the new Client implementation that is intended
     #   to become the new default in the next major release of this library.
     class UDPSink
-      def self.for_addr(addr)
+      def self.for_addr(addr, **kwargs)
         host, port_as_string = addr.split(":", 2)
-        new(host, Integer(port_as_string))
+        new(host, Integer(port_as_string), **kwargs)
       end
 
       attr_reader :host, :port
 
-      def initialize(host, port)
+      class << self
+        def finalizer(sink)
+          proc { sink.flush }
+        end
+      end
+
+      def initialize(host, port, batch_size: 1)
         @host = host
         @port = port
         @mutex = Mutex.new
         @socket = nil
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
+        @batch_size = batch_size
```

Comment:
Rather than batch size, I wonder whether we should instead (or in addition) measure the size of the full datagram that we are going to send. If this goes over MTU, there is a risk of the packet being dropped by the network.

The maximum safe size of a UDP packet is 508 bytes. However, in almost all cases, 1500 bytes is considered safe as well, as almost all networks are configured with an MTU around that size

So rather than building an array, does it make sense to build a string instead? This way we also prevent the `join` that would havew to iterate over the full array for every batch.

Reply:
So with the current implementation of simply sending once the buffer is full, a string would be too much trouble:

```ruby
if datagram.bytesize + buffer.bytesisze > MAX_BUFFER_SIZE
  flush
end
```

However if we go with a dispatcher thread (which I think we need otherwise as you noted there's no guarantee the datagram will be send anywhere in time), then it becomes harder. What do we do if the buffer is full? We have to store the datagram somewhere, and if we append to the string the dispatcher thread will have to do complex work to cut it back in pieces.

So just quickly like this I think we should continue with the array, but just be a bit smarter than a simple `join("\n")`. We can group the datagrams by packets of 500 bytes (or whatever the config is).